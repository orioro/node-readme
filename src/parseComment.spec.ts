import * as fs from 'fs'
import * as path from 'path'
import { parse as parse } from 'comment-parser/lib'
import {
  parseComment,
  parseCommentsFromFs
} from './parseComment'

const FIXTURES_PATH = path.join(__dirname, '../test')
const TMP_PATH = path.join(__dirname, '../tmp')

const FUNC1_COMMENT = parse(`
/**
 * Lorem ipsum dolor sit amet, consectetur adipiscing elit.
 * Quisque nec ligula sem.
 * 
 * @todo #65 Lorem ipsum dolor sit amet
 * @function func1
 * @param {String} param1 Param 1 description
 * @param {Number} [param2=0] Param 2 description
 * @returns {Boolean} result Result description
 */
`)[0]

const FUNC2_COMMENT = parse(`
/**
 * Lorem ipsum dolor sit amet, consectetur adipiscing elit.
 * Quisque nec ligula sem.
 * 
 * @function func2
 * @param {Object} param1 Param 1 description
 * @param {String} param1.key1 Key 1 description
 * @param {Boolean} param1.key2 Key 2 description
 * @param {Object[]} [param2=[]] Param 2 description
 * @param {Number} [param2[].key1]
 * @param {String} [param2[].key2]
 * @returns {Boolean} result Result description
 */
`)[0]

describe('parseComment(comment)', () => {
  test('@function / @func / @method', () => {

    expect(parseComment(FUNC1_COMMENT)).toMatchObject({
      name: 'func1',
      commentType: 'function',
      params: [
        { name: 'param1', type: 'String' },
        { name: 'param2', type: 'Number', optional: true, default: '0' }
      ],
      returns: { name: 'result', type: 'Boolean' },
      todo: [
        { name: '#65', description: 'Lorem ipsum dolor sit amet' }
      ]
    })
  })

  test('@function with nested param specs', () => {
    expect(parseComment(FUNC2_COMMENT)).toMatchObject({
      name: 'func2',
      commentType: 'function',
      params: [
        {
          name: 'param1',
          type: 'Object',
          properties: [
            { name: 'key1', type: 'String' },
            { name: 'key2', type: 'Boolean' },
          ]
        },
        {
          name: 'param2',
          type: 'Object[]',
          properties: [
            { name: 'key1', type: 'Number' },
            { name: 'key2', type: 'String' }
          ]
        }
      ],
      returns: { name: 'result', type: 'Boolean' }
    })
  })
})

describe('parseCommentsFromFs', () => {
  test('', () => {
    return parseCommentsFromFs('**/*.ts', {
      cwd: FIXTURES_PATH
    })
    .then(comments => {
      expect(comments).toMatchSnapshot()
    })
  })
})
