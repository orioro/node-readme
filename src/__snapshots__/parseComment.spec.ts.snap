// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseCommentsFromFs  1`] = `
Array [
  Object {
    "commentType": "typedef",
    "description": "Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris
nisi ut aliquip ex ea commodo consequat.",
    "examples": Array [],
    "file": "fixture-1/src/index.ts",
    "name": "CustomType",
    "params": Array [],
    "properties": Array [
      Object {
        "description": "",
        "name": "name",
        "optional": false,
        "problems": Array [],
        "properties": Array [],
        "source": Array [
          Object {
            "number": 9,
            "source": " * @property {String} name",
            "tokens": Object {
              "delimiter": "*",
              "description": "",
              "end": "",
              "name": "name",
              "postDelimiter": " ",
              "postName": "",
              "postTag": " ",
              "postType": " ",
              "start": " ",
              "tag": "@property",
              "type": "{String}",
            },
          },
        ],
        "tag": "property",
        "type": "String",
      },
      Object {
        "description": "",
        "name": "bool",
        "optional": false,
        "problems": Array [],
        "properties": Array [],
        "source": Array [
          Object {
            "number": 10,
            "source": " * @property {Boolean} bool",
            "tokens": Object {
              "delimiter": "*",
              "description": "",
              "end": "",
              "name": "bool",
              "postDelimiter": " ",
              "postName": "",
              "postTag": " ",
              "postType": " ",
              "start": " ",
              "tag": "@property",
              "type": "{Boolean}",
            },
          },
          Object {
            "number": 11,
            "source": " */",
            "tokens": Object {
              "delimiter": "",
              "description": "",
              "end": "*/",
              "name": "",
              "postDelimiter": "",
              "postName": "",
              "postTag": "",
              "postType": "",
              "start": " ",
              "tag": "",
              "type": "",
            },
          },
        ],
        "tag": "property",
        "type": "Boolean",
      },
    ],
    "todo": Array [
      Object {
        "description": "Some issue
",
        "name": "#3",
        "optional": false,
        "problems": Array [],
        "source": Array [
          Object {
            "number": 6,
            "source": " * @todo #3 Some issue",
            "tokens": Object {
              "delimiter": "*",
              "description": "Some issue",
              "end": "",
              "name": "#3",
              "postDelimiter": " ",
              "postName": " ",
              "postTag": " ",
              "postType": "",
              "start": " ",
              "tag": "@todo",
              "type": "",
            },
          },
          Object {
            "number": 7,
            "source": " * ",
            "tokens": Object {
              "delimiter": "*",
              "description": "",
              "end": "",
              "name": "",
              "postDelimiter": " ",
              "postName": "",
              "postTag": "",
              "postType": "",
              "start": " ",
              "tag": "",
              "type": "",
            },
          },
        ],
        "tag": "todo",
        "type": "",
      },
    ],
  },
  Object {
    "commentType": "function",
    "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit,
sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.

Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris
nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in
reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
pariatur.

\`\`\`js
const result = func1('some string', 10)
// false
\`\`\`

Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
    "examples": Array [],
    "file": "fixture-1/src/index.ts",
    "name": "func1",
    "params": Array [
      Object {
        "default": "''",
        "description": "Sed ut perspiciatis, unde omnis iste natus error
                         sit voluptatem accusantium doloremque laudantium.",
        "name": "strParam",
        "optional": true,
        "problems": Array [],
        "properties": Array [],
        "source": Array [
          Object {
            "number": 30,
            "source": " * @param {string} [strParam=''] Sed ut perspiciatis, unde omnis iste natus error",
            "tokens": Object {
              "delimiter": "*",
              "description": "Sed ut perspiciatis, unde omnis iste natus error",
              "end": "",
              "name": "[strParam='']",
              "postDelimiter": " ",
              "postName": " ",
              "postTag": " ",
              "postType": " ",
              "start": " ",
              "tag": "@param",
              "type": "{string}",
            },
          },
          Object {
            "number": 31,
            "source": " *                          sit voluptatem accusantium doloremque laudantium.",
            "tokens": Object {
              "delimiter": "*",
              "description": "sit voluptatem accusantium doloremque laudantium.",
              "end": "",
              "name": "",
              "postDelimiter": "                          ",
              "postName": "",
              "postTag": "",
              "postType": "",
              "start": " ",
              "tag": "",
              "type": "",
            },
          },
        ],
        "tag": "param",
        "type": "string",
      },
      Object {
        "default": "0",
        "description": "Nemo enim ipsam voluptatem, quia voluptas
                         sit, aspernatur aut odit aut fugit.",
        "name": "numParam",
        "optional": true,
        "problems": Array [],
        "properties": Array [],
        "source": Array [
          Object {
            "number": 32,
            "source": " * @param {number} [numParam=0] Nemo enim ipsam voluptatem, quia voluptas",
            "tokens": Object {
              "delimiter": "*",
              "description": "Nemo enim ipsam voluptatem, quia voluptas",
              "end": "",
              "name": "[numParam=0]",
              "postDelimiter": " ",
              "postName": " ",
              "postTag": " ",
              "postType": " ",
              "start": " ",
              "tag": "@param",
              "type": "{number}",
            },
          },
          Object {
            "number": 33,
            "source": " *                          sit, aspernatur aut odit aut fugit.",
            "tokens": Object {
              "delimiter": "*",
              "description": "sit, aspernatur aut odit aut fugit.",
              "end": "",
              "name": "",
              "postDelimiter": "                          ",
              "postName": "",
              "postTag": "",
              "postType": "",
              "start": " ",
              "tag": "",
              "type": "",
            },
          },
        ],
        "tag": "param",
        "type": "number",
      },
    ],
    "properties": Array [],
    "returns": Object {
      "description": "Nam libero tempore, cum soluta nobis est
                          eligendi optio.",
      "name": "result",
      "optional": false,
      "problems": Array [],
      "source": Array [
        Object {
          "number": 34,
          "source": " * @returns {boolean} result Nam libero tempore, cum soluta nobis est",
          "tokens": Object {
            "delimiter": "*",
            "description": "Nam libero tempore, cum soluta nobis est",
            "end": "",
            "name": "result",
            "postDelimiter": " ",
            "postName": " ",
            "postTag": " ",
            "postType": " ",
            "start": " ",
            "tag": "@returns",
            "type": "{boolean}",
          },
        },
        Object {
          "number": 35,
          "source": " *                           eligendi optio.",
          "tokens": Object {
            "delimiter": "*",
            "description": "eligendi optio.",
            "end": "",
            "name": "",
            "postDelimiter": "                           ",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 36,
          "source": " */",
          "tokens": Object {
            "delimiter": "",
            "description": "",
            "end": "*/",
            "name": "",
            "postDelimiter": "",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
      ],
      "tag": "returns",
      "type": "boolean",
    },
    "todo": Array [],
  },
  Object {
    "commentType": "function",
    "description": "Aliquam at metus molestie, luctus arcu id, ornare ligula.
View \`func1\`",
    "examples": Array [],
    "file": "fixture-1/src/index.ts",
    "name": "func2",
    "params": Array [
      Object {
        "description": "",
        "name": "param1",
        "optional": false,
        "problems": Array [],
        "properties": Array [
          Object {
            "description": "Some option",
            "name": "opt1",
            "optional": false,
            "problems": Array [],
            "source": Array [
              Object {
                "number": 48,
                "source": " * @param {string} param1.opt1 Some option",
                "tokens": Object {
                  "delimiter": "*",
                  "description": "Some option",
                  "end": "",
                  "name": "param1.opt1",
                  "postDelimiter": " ",
                  "postName": " ",
                  "postTag": " ",
                  "postType": " ",
                  "start": " ",
                  "tag": "@param",
                  "type": "{string}",
                },
              },
            ],
            "tag": "param",
            "type": "string",
          },
          Object {
            "description": "Some other option",
            "name": "opt2",
            "optional": false,
            "problems": Array [],
            "source": Array [
              Object {
                "number": 49,
                "source": " * @param {string} param1.opt2 Some other option",
                "tokens": Object {
                  "delimiter": "*",
                  "description": "Some other option",
                  "end": "",
                  "name": "param1.opt2",
                  "postDelimiter": " ",
                  "postName": " ",
                  "postTag": " ",
                  "postType": " ",
                  "start": " ",
                  "tag": "@param",
                  "type": "{string}",
                },
              },
            ],
            "tag": "param",
            "type": "string",
          },
        ],
        "source": Array [
          Object {
            "number": 47,
            "source": " * @param {Object} param1",
            "tokens": Object {
              "delimiter": "*",
              "description": "",
              "end": "",
              "name": "param1",
              "postDelimiter": " ",
              "postName": "",
              "postTag": " ",
              "postType": " ",
              "start": " ",
              "tag": "@param",
              "type": "{Object}",
            },
          },
        ],
        "tag": "param",
        "type": "Object",
      },
      Object {
        "description": "",
        "name": "param2",
        "optional": false,
        "problems": Array [],
        "properties": Array [],
        "source": Array [
          Object {
            "number": 50,
            "source": " * @param {number} param2",
            "tokens": Object {
              "delimiter": "*",
              "description": "",
              "end": "",
              "name": "param2",
              "postDelimiter": " ",
              "postName": "",
              "postTag": " ",
              "postType": " ",
              "start": " ",
              "tag": "@param",
              "type": "{number}",
            },
          },
        ],
        "tag": "param",
        "type": "number",
      },
    ],
    "properties": Array [],
    "returns": Object {
      "description": "",
      "name": "result",
      "optional": false,
      "problems": Array [],
      "source": Array [
        Object {
          "number": 51,
          "source": " * @returns {CustomType} result",
          "tokens": Object {
            "delimiter": "*",
            "description": "",
            "end": "",
            "name": "result",
            "postDelimiter": " ",
            "postName": "",
            "postTag": " ",
            "postType": " ",
            "start": " ",
            "tag": "@returns",
            "type": "{CustomType}",
          },
        },
        Object {
          "number": 52,
          "source": " */",
          "tokens": Object {
            "delimiter": "",
            "description": "",
            "end": "*/",
            "name": "",
            "postDelimiter": "",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
      ],
      "tag": "returns",
      "type": "CustomType",
    },
    "todo": Array [],
  },
  Object {
    "commentType": "function",
    "description": "",
    "examples": Array [],
    "file": "fixture-1/src/someMethod.ts",
    "name": "someMethod",
    "params": Array [
      Object {
        "description": "Excepteur sint occaecat cupidatat non proident",
        "name": "param",
        "optional": false,
        "problems": Array [],
        "properties": Array [],
        "source": Array [
          Object {
            "number": 8,
            "source": " * @param {number} param Excepteur sint occaecat cupidatat non proident",
            "tokens": Object {
              "delimiter": "*",
              "description": "Excepteur sint occaecat cupidatat non proident",
              "end": "",
              "name": "param",
              "postDelimiter": " ",
              "postName": " ",
              "postTag": " ",
              "postType": " ",
              "start": " ",
              "tag": "@param",
              "type": "{number}",
            },
          },
        ],
        "tag": "param",
        "type": "number",
      },
    ],
    "properties": Array [],
    "returns": Object {
      "description": "Lorem ipsum dolor sit amet, consectetur adipiscing",
      "name": "result",
      "optional": false,
      "problems": Array [],
      "source": Array [
        Object {
          "number": 9,
          "source": " * @returns {number} result Lorem ipsum dolor sit amet, consectetur adipiscing",
          "tokens": Object {
            "delimiter": "*",
            "description": "Lorem ipsum dolor sit amet, consectetur adipiscing",
            "end": "",
            "name": "result",
            "postDelimiter": " ",
            "postName": " ",
            "postTag": " ",
            "postType": " ",
            "start": " ",
            "tag": "@returns",
            "type": "{number}",
          },
        },
        Object {
          "number": 10,
          "source": " */",
          "tokens": Object {
            "delimiter": "",
            "description": "",
            "end": "*/",
            "name": "",
            "postDelimiter": "",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
      ],
      "tag": "returns",
      "type": "number",
    },
    "todo": Array [
      Object {
        "description": "Something to be done",
        "name": "-",
        "optional": false,
        "problems": Array [],
        "source": Array [
          Object {
            "number": 5,
            "source": " * @todo - Something to be done",
            "tokens": Object {
              "delimiter": "*",
              "description": "Something to be done",
              "end": "",
              "name": "-",
              "postDelimiter": " ",
              "postName": " ",
              "postTag": " ",
              "postType": "",
              "start": " ",
              "tag": "@todo",
              "type": "",
            },
          },
        ],
        "tag": "todo",
        "type": "",
      },
      Object {
        "description": "Something to be done that already has an issue number
",
        "name": "#45:OK",
        "optional": false,
        "problems": Array [],
        "source": Array [
          Object {
            "number": 6,
            "source": " * @todo #45:OK Something to be done that already has an issue number",
            "tokens": Object {
              "delimiter": "*",
              "description": "Something to be done that already has an issue number",
              "end": "",
              "name": "#45:OK",
              "postDelimiter": " ",
              "postName": " ",
              "postTag": " ",
              "postType": "",
              "start": " ",
              "tag": "@todo",
              "type": "",
            },
          },
          Object {
            "number": 7,
            "source": " *",
            "tokens": Object {
              "delimiter": "*",
              "description": "",
              "end": "",
              "name": "",
              "postDelimiter": "",
              "postName": "",
              "postTag": "",
              "postType": "",
              "start": " ",
              "tag": "",
              "type": "",
            },
          },
        ],
        "tag": "todo",
        "type": "",
      },
    ],
  },
]
`;
